### ISSUE TITLE
Windows icon with multiple sizes not used correctly

### ISSUE THREAD

**Issue by onelivesleft** [Nuitka/Nuitka#226]
The application icon is not generated correctly from the icon file; my icon file has multiple sizes in it, but nuitka only copies over the smallest one.  This means that it displays the icon correctly in explorer if the size is set to small, but when set to medium or larger (on the desktop for instance) it displays as if it had no icon.

Nuitka; 0.6.1
Python: 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:06:47) [MSC v.1914 32 bit (Intel)]
Executable: C:\python3\python.exe
OS: Windows
Arch: x86

Installed with pip

**Comment by kayhayen** [comment]
This is the resource part from `SingleExe.scons` that does this:

```
        if icon_path:
            rc_content.append(
                '2 ICON MOVEABLE PURE LOADONCALL DISCARDABLE "%s"' % (
                    icon_path.replace('\\', '/')
                )
            )
```

I have no clue what the attributes mean. Can you read up on the relevant documentation and find out what it could be?

Yours,
Kay

**Comment by onelivesleft** [comment]
I am similarly clueless about this, but I know that pyinstaller does it correctly, so perhaps you can use it as inspiration:
https://github.com/pyinstaller/pyinstaller/tree/1844d69f5aa1d64d3feca912ed1698664a3faf3e/PyInstaller

**Comment by kayhayen** [comment]
From the inspiring source, I get that `RT_GROUP_ICON` is a thing we don't do. Can you try changing `ICON` to `GROUP_ICON` in that `SingleExe.scons` of Nuitka. Maybe it's that easy.

PyInstaller seems to use Windows API to set the Icon after the binary has been built. Nuitka is using `rc.exe` from the Windows SDK, which will be more independent. For their code, they copy resources manually, but that's maybe not needed.

**Comment by kayhayen** [comment]
Recently I had added the ability to delete manifests from Windows DLLs, for which ctypes code to deal with resources as needed. The other day I expanded it to copy the manifest from the CPythonXY.DLL to created binaries, so they find the `MSCRT.DLL` configuration, so that copies resources.

These are in principle all steps needed to move the Icon handling out of the `rc.exe` handling and do it in Nuitka itself. Once that change stabilizes, I will make the icon resource copying manually and then add the multi icon support that way. My goal is to stop using not only `mt.exe` (used to extract manifest resources) but also `rc.exe` as this means also need for Windows SDK.

May take a while until I get there though, with GSoC taking up most of my time these days.

**Comment by kayhayen** [comment]
> I am similarly clueless about this, but I know that pyinstaller does it correctly, so perhaps you can use it as inspiration:
> https://github.com/pyinstaller/pyinstaller/blob/1844d69f5aa1d64d3feca912ed1698664a3faf3e/PyInstaller/utils/win32/icon.py

I think this is the way to go. We need not the full code there, basically just to parse and create the structure from the icon file. Then it would work. Not strictly rocket science. My problem is that I wouldn't even know how to test if it's correct. Is there something that would extract an icon again, so it can be compared, or an easy way to see all it's forms?

**Comment by onelivesleft** [comment]
I use this: https://icofx.ro/

**Comment by kayhayen** [comment]
So, Nuitka now allows to copy icon resources from existing files, as well as creating resources from multiple icon paths given. This should solve the issue. The code is currently on factory, but will be in the next release.

http://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
This is now part of the current pre-release and will be in 0.6.10

**Comment by kayhayen** [comment]
Released as part of 0.6.10
